File: 2024-08-August-GFG POTD/August 28 - Sorting Elements of an Array by Frequency/Solution.py
from collections import defaultdict

class Solution:
    def sortByFreq(self, arr):
        # Step 1: Count frequencies
        frequency_map = defaultdict(int)
        for num in arr:
            frequency_map[num] += 1
        
        # Step 2: Create a list of tuples
        freq_list = [(key, value) for key, value in frequency_map.items()]
        
        # Step 3: Sort the list
        freq_list.sort(key=lambda x: (-x[1], x[0])) # Sort by frequency (descending), then by number (ascending)
        
        # Step 4: Build the result
        result = []
        for pair in freq_list:
            result.extend([pair[0]] * pair[1])
        
        return result

if __name__ == "__main__":
    n = int(input())
    while n > 0:
        arr = list(map(int, input().split()))
        sol = Solution()
        ans = sol.sortByFreq(arr)
        print(" ".join(map(str, ans)))
        n -= 1

File: 2024-08-August-GFG POTD/August 28 - Sorting Elements of an Array by Frequency/Solution.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.lang.*;
import java.util.*;
import java.util.Map.Entry;


// } Driver Code Ends
// User function Template for Java

class Solution {
    // Function to sort the array according to frequency of elements.
    public ArrayList<Integer> sortByFreq(int arr[]) {
        // add your code here
        // Step 1: Count frequencies
        HashMap<Integer, Integer> frequencyMap = new HashMap<>();
        for (int num : arr) {
            frequencyMap.put(num, frequencyMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Create a list of pairs
        List<int[]> freqList = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : frequencyMap.entrySet()) {
            freqList.add(new int[]{entry.getKey(), entry.getValue()});
        }

        // Step 3: Sort the list
        Collections.sort(freqList, (a, b) -> {
            if (b[1] != a[1]) {
                return b[1] - a[1]; // Sort by frequency in descending order
            } else {
                return a[0] - b[0]; // If frequency is the same, sort by number in ascending order
            }
        });

        // Step 4: Build the result
        ArrayList<Integer> result = new ArrayList<>();
        for (int[] pair : freqList) {
            for (int i = 0; i < pair[1]; i++) {
                result.add(pair[0]);
            }
        }

        return result;
    }
}

//{ Driver Code Starts.

class Driverclass {
    public static void main(String[] args) throws NumberFormatException, IOException {
        BufferedReader sc = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(sc.readLine());
        while (n != 0) {
            String input = sc.readLine();
            String[] inputs = input.split(" ");
            int[] arr = new int[inputs.length];

            for (int i = 0; i < inputs.length; i++) {
                arr[i] = Integer.parseInt(inputs[i]);
            }

            ArrayList<Integer> ans = new ArrayList<Integer>();
            ans = new Solution().sortByFreq(arr);
            for (int i = 0; i < ans.size(); i++) System.out.print(ans.get(i) + " ");
            System.out.println();
            n--;
        }
    }
}

// } Driver Code Ends
